java-frontend/src/main/java/org/sonar/java/resolve/TypeSubstitutionSolver.java

```diff
@@ -74,17 +72,35 @@ JavaType getReturnType(@Nullable JavaType returnType, JavaType defSite, JavaType
     if (callSite != defSite) {
       resultType = applySiteSubstitution(resultType, callSite);
     }
-    if (isRawTypeOfParametrizedType(callSite) && !parametrizedMethodCall) {
-      // JLS8 5.1.9 + JLS8 15.12.2.6 : unchecked conversion
-      return resultType.erasure();
+    resultType = applySubstitution(resultType, substitution);
+    if(!isReturnTypeCompletelySubstituted(resultType, typeVariableTypes)) {
+      resultType = symbols.deferedType();
     }
-    return applySubstitution(resultType, substitution);
+    return resultType;
   }
 
-  private static boolean isRawTypeOfParametrizedType(JavaType site) {
-    return !isParametrizedType(site) && !site.symbol.typeVariableTypes.isEmpty();
+  private static boolean isReturnTypeCompletelySubstituted(JavaType resultType, List<TypeVariableJavaType> typeVariables){
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "isReturnTypeCompletelySubstituted" is 11 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)